File: sandbox-runner/patterns/batch-iterator/python/workflows.py
from datetime import timedelta

from temporalio import workflow
from temporalio.workflow import continue_as_new

from activities import fetch_page, process_record
from shared import PAGE_SIZE


@workflow.defn
class BatchIteratorWorkflow:
    """Processes PAGE_SIZE records per run, then calls continue_as_new with the
    next offset so history stays bounded."""

    @workflow.run
    async def run(self, offset: int = 0, total_processed: int = 0) -> int:
        page: list[int] = await workflow.execute_activity(
            fetch_page,
            args=[offset, PAGE_SIZE],
            start_to_close_timeout=timedelta(seconds=10),
        )

        for record_id in page:
            await workflow.execute_activity(
                process_record,
                record_id,
                start_to_close_timeout=timedelta(seconds=10),
            )
            total_processed += 1

        workflow.logger.info(
            f"Processed page: offset={offset} pageSize={len(page)} totalProcessed={total_processed}"
        )

        if len(page) == PAGE_SIZE:
            # More pages remain — continue as new with the next offset.
            continue_as_new(args=[offset + PAGE_SIZE, total_processed])

        # Reached here only on the final (partial) page.
        return total_processed

File: sandbox-runner/patterns/batch-iterator/typescript/workflows.ts
import { continueAsNew, log, proxyActivities } from "@temporalio/workflow";

import type * as activities from "./activities";
import { PAGE_SIZE } from "./shared";

const { fetchPage, processRecord } = proxyActivities<typeof activities>({
  startToCloseTimeout: "10 seconds",
});

/**
 * Batch Iterator Workflow: processes PAGE_SIZE records per run, then calls
 * continueAsNew with the next offset so history stays bounded.
 */
export async function batchIteratorWorkflow(
  offset: number = 0,
  totalProcessed: number = 0
): Promise<number> {
  const page = await fetchPage(offset, PAGE_SIZE);

  for (const recordId of page) {
    await processRecord(recordId);
    totalProcessed++;
  }

  log.info(`Processed page`, { offset, pageSize: page.length, totalProcessed });

  if (page.length === PAGE_SIZE) {
    // More pages remain — continue as new with the next offset.
    await continueAsNew<typeof batchIteratorWorkflow>(offset + PAGE_SIZE, totalProcessed);
  }

  // Reached here only on the final (partial) page.
  return totalProcessed;
}
